File: leetcode/1038-binary-search-tree-to-greater-sum-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstToGst(self, root: TreeNode) -> TreeNode:
        
        def dfs(node: TreeNode, curr_sum: int) -> int:
            if not node:
                return curr_sum
            right_sum = dfs(node.right, curr_sum)
            node.val = right_sum + node.val
            left_sum = dfs(node.left, node.val)
            return left_sum

        dfs(root, 0)
        return root

File: src/1038-binary-search-tree-to-greater-sum-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var bstToGst = function (root) {
    let sum = 0
    function helper(node) {
        //base
        if (node === null) return
        //recurtion

        helper(node.right)
        //work
        node.val = sum + node.val
        sum = node.val

        helper(node.left)
    }
    helper(root)
    return root

};